### PR TITLE
Fix destroyer task sometimes not being able to run

### LINKED ISSUE #10033: Deadlock scheduling finalizer task on failed build
Occurs in version range `5.5.1..5.7-20190721220031+0000`.

Repro projects:
- https://github.com/Triple-T/gradle-play-publisher/tree/96b63a8525b80ee4aa52a8072844e8783ad10be8
- https://github.com/SUPERCILEX/issue-mcve/tree/52ae3dd651c054495bcd2f5e0f74e5390ec0fb55

Repro steps:
1. In `gradle-play-publisher`: `./gradlew publishToMavenLocal`
1. In `issue-mcve`: `./gradlew pApk`

`publishReleaseApk` will fail (as expected), but the **build will now be frozen forever**. This is because there's a finalizer task. Somehow, Gradle is deadlocking when trying to schedule the finalizer task.

### PR DESCRIPTION
**Pull request by wolfs**
Tasks which are skipped should not block other tasks from running.

Fixes #10033

### CODE REVIEW

File: subprojects/core/src/main/java/org/gradle/execution/plan/DefaultExecutionPlan.java [hunk 1]
```diff
@@ -611,7 +611,7 @@ public Node selectNext(WorkerLeaseRegistry.WorkerLease workerLease, ResourceLock
                     node.startExecution();
                 } else {
                     node.skipExecution();
-                    updateAllDependenciesCompleteForPredecessors(node);
+                    recordNodeCompleted(node);
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/execution/plan/DefaultExecutionPlan.java]: I wonder if we could do something like `Node.completeNode(Consumer<Node>)` so we could do this:

```java
completeNode(node, Node::skipExecution);
```

What I'm trying to achieve here is to remove the risk of calling `skipExecution()` etc. without completing the node.

Same with `recordNodeStarted()`, though that's only called from one place, so it's more about consistency. Another solution would be to rename `recordNodeStarted()` to `startNode()` and let it call `node.startExecution()`.

File: subprojects/core/src/main/java/org/gradle/execution/plan/DefaultExecutionPlan.java [hunk 2]
```diff
@@ -931,7 +931,6 @@ private void recordNodeStarted(Node node) {
     }
 
     private void recordNodeCompleted(Node node) {
-        runningNodes.remove(node);
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/execution/plan/DefaultExecutionPlan.java]: What's the difference between `recordNodeCompleted()` and `nodeComplete()`? Can we clarify the difference a bit more by a different name maybe?

File: subprojects/core/src/main/java/org/gradle/execution/plan/DefaultExecutionPlan.java (hunk 1 above)
Comment (lptr) [reply subprojects/core/src/main/java/org/gradle/execution/plan/DefaultExecutionPlan.java]: `aborted` could also be set by `completeNode()` explicitly.

File: subprojects/core/src/main/java/org/gradle/execution/plan/DefaultExecutionPlan.java (hunk 2 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/execution/plan/DefaultExecutionPlan.java]: `nodeCompleted == node finished executing` which means that the node also executed. `recordNodeCompleted` is more about doing the bookkeeping when the node moves to the completed state.

Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/execution/plan/DefaultExecutionPlan.java]: I renamed `nodeCompleted` to `finishedExecuting`.

File: subprojects/core/src/main/java/org/gradle/execution/plan/DefaultExecutionPlan.java (hunk 1 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/execution/plan/DefaultExecutionPlan.java]: The completion and starting methods in `Node` now take an extra consumer for the bookkeeping.

**Comment by wolfs** [comment]
@lptr PTAL!

**Pull request merged**